src/component.js

```diff
@@ -24,68 +24,59 @@ const excludedProps = NOT_KEY_PATH_PROPS.reduce((val, prop) => {
 }, {})
 
 export default class TideComponent extends React.Component {
-  constructor(props) {
-    super(...arguments)
-    const tide = this.getTide()
-    const keyPaths = this.getKeyPaths(props, tide)
-    this._componentTide = {
-      keyPaths,
-      ...tide.getComponentProps(),
-    }
-
-    this.state = this.getPropsFromKeyPaths(keyPaths, tide)
+  constructor(...args) {
+    super(...args)
+    this.tide = this.props.tide instanceof Tide ? this.props.tide : this.context.tide
+    this.state = this.getMappedProps()
   }
 
   getChildContext() {
-    return {tide: this.getTide()}
+    return {tide: this.tide}
   }
 
   componentWillMount() {
-    this._onStateChange = this.onStateChange.bind(this)
-    this.getTide().onChange(this._onStateChange)
+    this.tide.onChange(this.handleTideStateChange)
   }
 
   componentWillUnmount() {
     this._isUnmounting = true
-    this.getTide().offChange(this._onStateChange)
+    this.tide.offChange(this.handleTideStateChange)
   }
 
-  onStateChange() {
+  handleTideStateChange = () => {
     if (this._isUnmounting) return
-    const tide = this.getTide()
-    const newState = this.getPropsFromKeyPaths(this.getKeyPaths(this.props, tide), tide)
-    if (this.hasStaleState(newState)) {
+
+    const newState = this.getMappedProps()
+    if (shallowEqual(newState, this.state) === false) {
       this.setState(newState)
     }
   }
 
-  getKeyPaths(props, tide) {
-    let keyPaths = omit(props, (key) => excludedProps[key])
+  getKeyPaths() {
+    let keyPaths = omit(this.props, (key) => excludedProps[key])
+
     keyPaths = mapValues(keyPaths, (val, key) => {
-      const value = typeof val === 'function' ? val(tide.getState()) : val
+      const value = typeof val === 'function' ? val(this.tide.getState()) : val
       if (Array.isArray(value)) return value
       if (value === true) return [key]
       return value.split('.')
     })
     return keyPaths
   }
 
-  getPropsFromKeyPaths(keyPaths, tide) {
-    const state = tide.getState()
-    return mapValues(keyPaths, (value) => state.getIn(value))
-  }
-
-  getTide() {
-    return this.props.tide instanceof Tide ?
-      this.props.tide : this.context.tide
+  getMappedProps() {
+    const state = this.tide.getState()
+    return mapValues(this.getKeyPaths(), (value) => state.getIn(value))
   }
 
   getChildProps() {
-    return {...omit(this.state, (_, value) => value === undefined), tide: this._componentTide}
-  }
+    const tide = {
+      keyPaths: this.getKeyPaths(),
+      ...this.tide.getComponentProps(),
+    }
```

Comment:
Sure, check out the benchmarking here: https://gist.github.com/lwakefield/8e8a702dd9b8df7679174e468bad8706

The refactor is a little slower (~3%). Let me see what I can do, but I think we can do better than the existing solution!